cantools/database/diagnostics/data.py

```diff
@@ -59,35 +67,69 @@ def __init__(
         #: if unavailable.
         self.choices: Optional[Choices] = choices
 
+        #: This stores start and end points of piecewise linear segments
+        #  None for all other types
+        self.pcw_segments: _SEGMENTLISTLIMITTYPE = None
+        if pcw_segments is not None:
+            self._initialize_segment_limits(pcw_segments)
+
         # ToDo: Remove once types are handled properly.
         self.is_float: bool = False
         self.is_signed: bool = False
 
-    def get_offset_scaling(self, raw_val):
+    def _initialize_segment_limits(self, pcw_segments: List[Tuple[float, float]]) -> None:
         def convert(v, o, f):
-            return (v - o) / f
-
-        if not isinstance(self.minimum, list):
-            return self.offset, self.scale
+            return v * f + o
+
+        self.pcw_segments = []
+        last_phys_max = None
+        for i, segment in enumerate(pcw_segments):
+            start, end = segment
+            params = [self.offset[i], self.scale[i]]
+            parsed_segment: _SEGMENTDICTDEF = {
+                "raw": (start, end),
+                "scaled": (convert(start, *params), convert(end, *params))
+            }
+            self.pcw_segments.append(parsed_segment)
+            if last_phys_max is None:
+                last_phys_max = parsed_segment["scaled"][1]
+                continue
+
+            if last_phys_max >= parsed_segment["scaled"][0]:
+                logger.warning(f"Piecewise linear type: {self.name} has overlapping segments! "
+                               f"Segment {i} starts at phys val {parsed_segment['scaled'][0]} "
+                               f"but one of the prev segments ended at {last_phys_max}. "
+                               "Encoding might be ambiguous.")
+
+            last_phys_max = max(parsed_segment["scaled"][1], last_phys_max)
+
+    def get_offset_scaling(self, raw_val: Optional[float] = None, scaled: Optional[float] = None) -> Tuple[float, float]:
+        if raw_val is None and scaled is None:
+            raise ValueError("Either raw or scaled value needs to be given!")
+
+        if self.pcw_segments is None or len(self.pcw_segments) == 0:
+            return self.offset[0], self.scale[0]
 
-        for i in range(len(self.minimum)):
-            if self.minimum[i] <= raw_val <= self.maximum[i]:
-                return self.offset[i], self.scale[i]
-            elif (
-                self.minimum[i]
-                <= convert(raw_val, self.offset[i], self.scale[i])
-                <= self.maximum[i]
-            ):
+        type_specifer: Literal["raw", "scaled"] = "raw"
+        val = raw_val
+        if raw_val is None:
+            val = scaled
+            type_specifer = "scaled"
```

Comment:
this seems to be a rather complicated way of doing things. better split the function...